source/common/network/socket_impl.cc

```diff
@@ -134,5 +134,53 @@ Address::InstanceConstSharedPtr SocketInterface::peerAddressFromFd(os_fd_t fd) {
   return Address::addressFromSockAddr(ss, ss_len);
 }
 
+SocketImpl::SocketImpl(Address::SocketType type, Address::Type addr_type,
+                       Address::IpVersion version)
+    : io_handle_(SocketInterface::socket(type, addr_type, version)) {}
+
+SocketImpl::SocketImpl(Address::SocketType socket_type, const Address::InstanceConstSharedPtr addr)
+    : io_handle_(SocketInterface::socket(socket_type, addr)) {}
+
+Api::SysCallIntResult SocketImpl::bind(Network::Address::InstanceConstSharedPtr address) {
+  const auto* address_base = dynamic_cast<const Network::Address::InstanceBase*>(address.get());
```

Comment:
The dynamic casts in this code are somewhat of a code/interface smell, as we now have a tight binding between the two interface implementations, which I think is not what we want. Can we brainstorm ways of breaking this dependency? I think the issue is you need access to the unix sockaddr? I'm not sure of the right solution here, perhaps some type of visitor that can prep a socket (though that gets us back where we started), or possibly somehow exposing a generic/unix view of the address that can be then converted somehow?

Reply:
One option is adding `sockAddr()` to the interface, instead of only to an implementation. Do we think we'll have an implementation of Socket that requires something other than `struct sockaddr`?

Reply:
Given that sockaddr is effectively just the address bytes, I think other platforms could do a conversion so IMO this seems fine to me.

Reply:
Adding sockaddr to the Address::Instance interface, but for pipes we still need to grab the abstract_namespace and mode. So, as far as I can tell, we'll still need to downcast in that case, but I guess that's fine?

Reply:
I would rather not downcast in any case. Can we just add the name/namespace/mode to the interface as well for now? We can work out conversions later if needed.

Reply:
I don't think you need an accessor for `abstract_namespace_`. You can check the first byte of `sun_path` for `\0` to tell. But you will need `mode` I think. I would add a new address interface type for pipe addresses with an accessor for `mode()`, and add an accessor that type on `Address::Instance` similar to `ip()`. Then there's no need to downcast.

Reply:
@ggreenway I uploaded the patch before reading this. I think I did exactly what you suggested but also added an abstractNamespace accessor. Let me know if you want me to take it out